### ISSUE TITLE
[2.0.0 release] Bug report - Failed to list nettests error_code: 400 backend_message: _director_started must be True

### ISSUE THREAD

**Issue by anadahz** [TheTorProject/ooni-probe#601]
`# uptime`
`12:57:08 up  9:47,  1 user,  load average: 0.01, 0.01, 0.00`

ooniprobe v2.0.0-rc.1
---------------------------
After fresh start of ooniprobe and by selecting the default deck in WUI:

[decks]
No decks are being listed.
Error message: `Failed to list nettests error_code: 400 backend_message: _director_started must be True`

[nettests]
`Warning ooniprobe has not started yet. Please wait for it to start running tests.`

`$ systemctl status ooniprobe -l`
```
● ooniprobe.service - ooniprobe.service, network interference detection tool
   Loaded: loaded (/etc/systemd/system/ooniprobe.service; enabled)
   Active: active (running) since Thu 1970-01-01 00:00:42 UTC; 46 years 8 months ago
  Process: 327 ExecStart=/usr/local/bin/ooniprobe-agent start (code=exited, status=0/SUCCESS)
 Main PID: 582 (ooniprobe-agent)
   CGroup: /system.slice/ooniprobe.service
           └─582 /usr/bin/python /usr/local/bin/ooniprobe-agent start
```
Output of `/var/log/ooniprobe.log`:
```
1970-01-01 00:00:42+0000 [-] [!] Failed to lookup via https://api.duckduckgo.com/?q=ip&format=json
1970-01-01 00:00:42+0000 [-] Unhandled Error
	Traceback (most recent call last):
	Failure: twisted.internet.error.DNSLookupError: DNS lookup failed: address 'api.duckduckgo.com' not found: [Errno -2] Name or service not known.
	
1970-01-01 00:00:42+0000 [-] [!] Failed to lookup via http://geoip.ubuntu.com/lookup
1970-01-01 00:00:42+0000 [-] Unhandled Error
	Traceback (most recent call last):
	Failure: twisted.internet.error.DNSLookupError: DNS lookup failed: address 'geoip.ubuntu.com' not found: [Errno -2] Name or service not known.
	
1970-01-01 00:00:42+0000 [-] [!] Failed to run update-inputs
1970-01-01 00:00:42+0000 [-] Unhandled Error
	Traceback (most recent call last):
	  File "/usr/local/lib/python2.7/dist-packages/twisted/internet/defer.py", line 434, in errback
	    self._startRunCallbacks(fail)
	  File "/usr/local/lib/python2.7/dist-packages/twisted/internet/defer.py", line 501, in _startRunCallbacks
	    self._runCallbacks()
	  File "/usr/local/lib/python2.7/dist-packages/twisted/internet/defer.py", line 587, in _runCallbacks
	    current.result = callback(current.result, *args, **kw)
	  File "/usr/local/lib/python2.7/dist-packages/twisted/internet/defer.py", line 1241, in gotResult
	    _inlineCallbacks(r, g, deferred)
	--- <exception caught here> ---
	  File "/usr/local/lib/python2.7/dist-packages/twisted/internet/defer.py", line 1183, in _inlineCallbacks
	    result = result.throwExceptionIntoGenerator(g)
	  File "/usr/local/lib/python2.7/dist-packages/twisted/python/failure.py", line 389, in throwExceptionIntoGenerator
	    return g.throw(self.type, self.value, self.tb)
	  File "/usr/local/lib/python2.7/dist-packages/ooniprobe-2.0.0rc1-py2.7.egg/ooni/agent/scheduler.py", line 112, in run
	    yield self.task()
	  File "/usr/local/lib/python2.7/dist-packages/twisted/internet/defer.py", line 1183, in _inlineCallbacks
	    result = result.throwExceptionIntoGenerator(g)
	  File "/usr/local/lib/python2.7/dist-packages/twisted/python/failure.py", line 389, in throwExceptionIntoGenerator
	    return g.throw(self.type, self.value, self.tb)
	  File "/usr/local/lib/python2.7/dist-packages/ooniprobe-2.0.0rc1-py2.7.egg/ooni/agent/scheduler.py", line 127, in task
	    yield probe_ip.lookup()
	  File "/usr/local/lib/python2.7/dist-packages/twisted/internet/defer.py", line 1183, in _inlineCallbacks
	    result = result.throwExceptionIntoGenerator(g)
	  File "/usr/local/lib/python2.7/dist-packages/twisted/python/failure.py", line 389, in throwExceptionIntoGenerator
	    return g.throw(self.type, self.value, self.tb)
	  File "/usr/local/lib/python2.7/dist-packages/ooniprobe-2.0.0rc1-py2.7.egg/ooni/geoip.py", line 225, in lookup
	    yield self.askGeoIPService()
	  File "/usr/local/lib/python2.7/dist-packages/twisted/internet/defer.py", line 1183, in _inlineCallbacks
	    result = result.throwExceptionIntoGenerator(g)
	  File "/usr/local/lib/python2.7/dist-packages/twisted/python/failure.py", line 389, in throwExceptionIntoGenerator
	    return g.throw(self.type, self.value, self.tb)
	  File "/usr/local/lib/python2.7/dist-packages/ooniprobe-2.0.0rc1-py2.7.egg/ooni/geoip.py", line 250, in askGeoIPService
	    raise errors.ProbeIPUnknown
	ooni.errors.ProbeIPUnknown: 
```

**Comment by hellais** [comment]
It seems like there is a network connectivity issue on the machine running this at least in DNS resolution as demonstrated by:

```
    Failure: twisted.internet.error.DNSLookupError: DNS lookup failed: address 'geoip.ubuntu.com' not found: [Errno -2] Name or service not known.
    Failure: twisted.internet.error.DNSLookupError: DNS lookup failed: address 'geoip.ubuntu.com' not found: [Errno -2] Name or service not known.
```

**Comment by anadahz** [comment]
`/var/log/tor/log` excerpt:
```
Jan 01 00:00:50.000 [warn] 5 connections have failed:
Jan 01 00:00:50.000 [warn]  5 connections died in state connect()ing with SSL state (No SSL object)
Sep 08 03:17:42.000 [warn] comparing times on millisecond detail too far apart: 1473304199 seconds
Sep 08 03:17:43.000 [notice] Your system clock just jumped 1473304200 seconds forward; assuming established circuits no longer work.
Sep 08 03:17:44.000 [notice] Bootstrapped 10%: Finishing handshake with directory server
Sep 08 03:20:54.000 [notice] Your system clock just jumped 189 seconds forward; assuming established circuits no longer work.
Sep 08 04:10:31.000 [notice] Bootstrapped 15%: Establishing an encrypted directory connection
Sep 08 04:10:31.000 [notice] Bootstrapped 20%: Asking for networkstatus consensus
Sep 08 04:10:31.000 [notice] Bootstrapped 25%: Loading networkstatus consensus
Sep 08 04:10:36.000 [notice] I learned some more directory information, but not enough to build a circuit: We have no usable consensus.
Sep 08 04:10:38.000 [notice] Bootstrapped 40%: Loading authority key certs
Sep 08 04:10:40.000 [notice] Bootstrapped 45%: Asking for relay descriptors
Sep 08 04:10:40.000 [notice] I learned some more directory information, but not enough to build a circuit: We need more microdescriptors: we have 0/7092, and can only build 0% of likely paths. (We have 0% of guards bw, 0% of midpoint bw, and 0% of exit bw = 0% of path bw.)
Sep 08 04:10:42.000 [notice] Bootstrapped 50%: Loading relay descriptors
Sep 08 04:10:58.000 [notice] Bootstrapped 56%: Loading relay descriptors
Sep 08 04:10:59.000 [notice] Bootstrapped 62%: Loading relay descriptors
Sep 08 04:11:03.000 [notice] Bootstrapped 73%: Loading relay descriptors
Sep 08 04:11:04.000 [notice] Bootstrapped 80%: Connecting to the Tor network
Sep 08 04:11:05.000 [notice] Bootstrapped 90%: Establishing a Tor circuit
Sep 08 04:11:09.000 [notice] Tor has successfully opened a circuit. Looks like client functionality is working.
Sep 08 04:11:09.000 [notice] Bootstrapped 100%: Done
Sep 08 09:10:28.000 [notice] Heartbeat: Tor's uptime is 5:56 hours, with 0 circuits open. I've sent 814 kB and received 4.01 MB.
```

**Comment by anadahz** [comment]
`# host duckduckgo.com`
```
duckduckgo.com has address 184.72.106.52
duckduckgo.com has address 54.208.100.124
duckduckgo.com has address 107.21.1.61
duckduckgo.com has address 184.72.115.86
duckduckgo.com mail is handled by 20 in2-smtp.messagingengine.com.
duckduckgo.com mail is handled by 50 mx3.dnsmadeeasy.com.
duckduckgo.com mail is handled by 10 in1-smtp.messagingengine.com.
duckduckgo.com mail is handled by 60 mx2.dnsmadeeasy.com.
duckduckgo.com mail is handled by 40 mx1.dnsmadeeasy.com.
```

**Comment by hellais** [comment]
I suspect this is related to the clock being set to 1970 and due to that networking not working properly. I implemented in https://github.com/TheTorProject/ooni-probe/commit/4f80a3bb07df8d742cb79f8ca918734412159422 a retry system for the director startup with exponential backoff.

This should be enough to overcome this sort of issue in most cases.

I have updated the rc1 pushed to https://people.torproject.org/~art/ooni/ooniprobe-2.0.0rc1.tar.gz with this commit. If you have a way of reproducing this environment and checking if it solves it that would be good.
